src/v/serde/avro/tests/parser_test.cc

```diff
@@ -0,0 +1,469 @@
+                }
+            }
+            return AssertionSuccess();
+        }
+        case avro::AVRO_MAP: {
+            auto expected_map = expected.value<::avro::GenericMap>();
+            auto actual_map = actual.value<::avro::GenericMap>();
+            if (expected_map.value().size() != actual_map.value().size()) {
+                return AssertionFailure() << fmt::format(
+                         "map size mismatch {} != {}",
+                         expected_map.value().size(),
+                         actual_map.value().size());
+            }
+            for (size_t i : boost::irange(expected_map.value().size())) {
+                if (
+                  expected_map.value()[i].first
+                  != actual_map.value()[i].first) {
+                    return AssertionFailure() << fmt::format(
+                             "map key mismatch {} != {}",
+                             expected_map.value()[i].first,
+                             actual_map.value()[i].first);
+                }
+                auto result = are_equal(
+                  expected_map.value()[i].second, actual_map.value()[i].second);
+                if (!result) {
+                    return AssertionFailure() << fmt::format(
+                             "map value mismatch - {}", result.message());
+                }
+            }
+            return AssertionSuccess();
+        }
+
+        case avro::AVRO_UNION:
+            if (expected.unionBranch() != actual.unionBranch()) {
+                return AssertionFailure() << fmt::format(
+                         "union branch mismatch {} != {}",
+                         expected.unionBranch(),
+                         actual.unionBranch());
+            }
+            return are_equal(
+              expected.value<::avro::GenericUnion>().datum(),
+              actual.value<::avro::GenericUnion>().datum());
+        case avro::AVRO_FIXED: {
+            auto expected_fixed = expected.value<::avro::GenericFixed>();
+            auto actual_fixed = expected.value<::avro::GenericFixed>();
+            if (expected_fixed.value() != actual_fixed.value()) {
+                return AssertionFailure()
+                       << fmt::format("fixed value mismatch");
+            }
+            return AssertionSuccess();
+        }
+        case avro::AVRO_UNKNOWN:
+        case avro::AVRO_NUM_TYPES:
+            break;
+        }
+        return AssertionFailure();
+    }
+};
+
+// Basic AVRO parsing test. The idea is very simple, the test do the following
+// steps:
+// - loads JSON schema from a file
+// - generates random GenericDatum based on the loaded schema
+// - serializes the data using avro::avro_write (generic encoding)
+// - deserializes data using `serde::avro::parse`
+// - converts deserialized data back to GenericDatum
+// - validates if the original value is equal to the one which was deserialized
+TEST_P(AvroParserTest, RoundtripTest) {
```

Comment:
Love the tests - nice work!

Should we also add some tests for failure cases? IE empty input, input too long, using an invalid schema to make sure we handle corrupt data correctly?

Additionally would be a nice bonus to have a testcase for max nesting depth - you could probably make the generator take a parameter to generate a different depth to trigger that case.

Reply:
there is a problem with handling incorrect data, especially when it comes to the size of arrays, as we decode varlong we can always decode the wrong value leading to oversized allocation, should we have safeguards here ?

Reply:
Yes totally! That is a good callout. I think just ensuring the resulting array's length is not greater than remaining bytes is good. I will check if I need to update the protobuf parser

Reply:
Looks like the protobuf parser is OK. Note that I don't think reserve will cause any alloc problems on a chunked_vector. Reserve only works on the first segment, but we should still add validation (we do this in the protobuf parser).

Reply:
https://github.com/redpanda-data/redpanda/blob/1754e471e487e66f568e5128b61d1261965cefe5/src/v/serde/protobuf/parser.cc#L477-L483